Add Challenges 20 21 22 for str Pattern and iter (#266)
model-checking/verify-rust-std: doc/src/challenges/0021-str-pattern-pt2.md

> # Challenge 21: Verify the correctness of char-related functions in str::pattern

Reviewer: Do we want to do correctness at this time?

Author: Sorry, it is a typo. I will edit.

1. The safety and functional correctness of all functions in `slice` module. 
2. That all functions in (library/core/src/str/validations.rs) are functionally correct (consistent with the UTF8 encoding description in https://en.wikipedia.org/wiki/UTF-8). 
3. That all the Searchers in (library/core/src/str/iter.rs) are created by the into_searcher(_, haystack) with haystack is a valid utf8 string (str). You can assume any utf8 string property of haystack.

Verify the safety of the functions in (library/core/src/str/pattern.rs) listed in the next section.

The safety properties we are targeting are: 
1. There is no UB happens when calling the functions after the Searcher is created.
2. The impls of unsafe traits `Searcher` and `ReverseSearcher` satisfies the SAFETY condition stated in the file: 
```
/// The trait is marked unsafe because the indices returned by the
/// [`next()`][Searcher::next] methods are required to lie on valid utf8
/// boundaries in the haystack. This enables consumers of this trait to
/// slice the haystack without additional runtime checks.
```
This property should hold for next_back() of `ReverseSearcher` too.

### Success Criteria

> Verify the safety of the following functions in (library/core/src/str/pattern.rs) : next, next_match, next_back, next_match_back, next_reject, next_back_reject

Reviewer 2: Let's just combine this with the previous challenge--if one of them gets solved first, it'll be trivial for someone to do the other, so it doesn't seem fair to have them be independent and have the same reward. If we think it's a harder challenge combined, we can up the reward.

Author: I don't think we should combine the two this time. The implementation of StrSearcher (for substring) is very different from (and more complicated than) those of Searchers for char or char-slice.

# Challenge 21: Verify the safety of substring-related functions in str::pattern

- **Status:** Open
- **Tracking Issue:** [#278](https://github.com/model-checking/verify-rust-std/issues/278)
- **Start date:** *2025-03-07*
- **End date:** *2025-10-17*
- **Reward:** *5000 USD*

-------------------

> ### Context

Reviewer 2: Can you apply the changes I requested in the previous challenge to the context and details sections here too?

Reviewer 2: I think we should differentiate this opening a bit from the previous challenge--can you tailor it more to `StrSearcher`, and how substrings are a different challenge? I believe you that these should be separate challenges, but I think the fact that the beginnings of the challenges are identical implies that they're more similar than they are.

Reviewer 2: (FWIW, I think it would be fine to delete this details and context entirely, refer people to the previous challenge for those details, and then write 1-2 sentences about how we're going to do the same kind of verification for substrings here).

Author: I added 2 sentences at the end of the Details section.

1. The safety and functional correctness of all functions in `slice` module. 
2. That all functions in (library/core/src/str/validations.rs) are functionally correct (consistent with the UTF8 encoding description in https://en.wikipedia.org/wiki/UTF-8). 
3. That all the Searchers in (library/core/src/str/iter.rs) are created by the into_searcher(_, haystack) with haystack is a valid utf8 string (str). You can assume any utf8 string property of haystack.

Verify the safety of the functions in (library/core/src/str/pattern.rs) listed in the next section.

The safety properties we are targeting are: 
1. There is no UB happens when calling the functions after the Searcher is created.
2. The impls of unsafe traits `Searcher` and `ReverseSearcher` satisfies the SAFETY condition stated in the file: 
```
/// The trait is marked unsafe because the indices returned by the
/// [`next()`][Searcher::next] methods are required to lie on valid utf8
/// boundaries in the haystack. This enables consumers of this trait to
/// slice the haystack without additional runtime checks.
```
This property should hold for next_back() of `ReverseSearcher` too.

### Success Criteria

> Verify the safety of the following functions in (library/core/src/str/pattern.rs) : next, next_match, next_back, next_match_back, next_reject, next_back_reject
> which are implemented for `StrSearcher`.

Reviewer 2: ```suggestion
Verify the safety of the following `StrSearcher` functions in (library/core/src/str/pattern.rs): 
- `next`
- `next_match`
- `next_back`
- `next_match_back`
- `next_reject`
- `next_back_reject`
```

> The verification is consider successful if you can specify a condition (a "type invariant") C and prove that:
> 1. If the `StrSearcher` is created from any valid utf8 haystack, it satisfies C.
> 2. If the `StrSearcher` satisfies C, it ensures the two safety properties mentioned in the previous section.
> 3. If the `StrSearcher` satisfies C, after it calls any function above and gets modified, it still statisfies C.

Reviewer 2: ```suggestion
The verification is considered successful if you can specify a condition (a "type invariant") `C` and prove that:
1. If the `StrSearcher` is created from any valid UTF-8 haystack, it satisfies `C`.
2. If the `StrSearcher` satisfies `C`, it ensures the two safety properties mentioned in the previous section.
3. If the `StrSearcher` satisfies `C`, after it calls any function above and gets modified, it still satisfies `C`.
```

- **End date:** *2025-10-17*
- **Reward:** *25000 USD*

-------------------
## Goal
Verify the safety of `StrSearcher` implementation in `str::pattern`.

## Motivation

String and `str` types are widely used in Rust programs, so it is important that their associated functions do not cause undefined behavior.

## Description

The following str library functions are generic over the `Pattern` trait (https://doc.rust-lang.org/std/str/pattern/trait.Pattern.html): 
- `contains`
- `starts_with`
- `ends_with`
- `find` 
- `rfind`
- `split`
- `split_inclusive`
- `rsplit`
- `split_terminator`
- `rsplit_terminator`
- `splitn`
- `rsplitn`
- `split_once`
- `rsplit_once`
- `rmatches`
- `match_indices`
- `rmatch_indices`
- `trim_matches`
- `trim_start_matches`
- `strip_prefix`
- `strip_suffix`
- `trim_end_matches`
These functions accept a pattern as input, then call [into_searcher](https://doc.rust-lang.org/std/str/pattern/trait.Pattern.html#tymethod.into_searcher) to create a [Searcher](https://doc.rust-lang.org/std/str/pattern/trait.Pattern.html#associatedtype.Searcher) for the pattern. They use this `Searcher` to perform their desired operations (split, find, etc.).

> Those functions are implemented in (library/core/src/str/mod.rs), but the core of them is the searching algorithms which are implemented in (library/core/src/str/pattern.rs).

Reviewer: ```suggestion
Those functions are implemented in (library/core/src/str/mod.rs), but the core of them are the searching algorithms which are implemented in (library/core/src/str/pattern.rs).
```

- **Tracking Issue:** [#278](https://github.com/model-checking/verify-rust-std/issues/278)
- **Start date:** *2025-03-07*
- **End date:** *2025-10-17*
- **Reward:** *10000 USD*

-------------------

> 

Reviewer 2: ```suggestion
## Goal
Verify the safety of `StrSearcher` methods in `str::pattern`.

```

I think having this goals sentence in the beginning is enough to distinguish between this challenge and the previous one at a glance, and that's all I really wanted.

- **Reward:** *10000 USD*

-------------------
## Goal
Verify the safety of `StrSearcher` implementation in `str::pattern`.

## Motivation

> String and str types are widely used in Rust programs. Verifying Rust String and str functions in Rust standard library is important in ensuring the safety of these programs.

Reviewer 2: ```suggestion
String and `str` types are widely used in Rust programs, so it is important that their associated functions do not cause undefined behavior.
```

These functions accept a pattern as input, then call [into_searcher](https://doc.rust-lang.org/std/str/pattern/trait.Pattern.html#tymethod.into_searcher) to create a [Searcher](https://doc.rust-lang.org/std/str/pattern/trait.Pattern.html#associatedtype.Searcher) for the pattern. They use this `Searcher` to perform their desired operations (split, find, etc.).
Those functions are implemented in (library/core/src/str/mod.rs), but the core of them is the searching algorithms which are implemented in (library/core/src/str/pattern.rs).

> ### Assumption

Reviewer 2: ```suggestion
### Assumptions
```

```
This property should hold for next_back() of `ReverseSearcher` too.

> Although this challenge appears similar to Challenge 20 and only requires proving safety for the `StrSearcher` implementation, 
> it warrants being a separate challenge due to its complexity. 
> The `StrSearcher` implementation relies on sophisticated algorithms - TwoWay-Search and SIMD-search - making the safety proof significantly more complex.

Reviewer 2: ```suggestion
```

I think this is confusing to have here because this is the first mention of `StrSearcher`--we should tell people what the challenge is about before justifying the choice.

Reviewer 2: My request was more to just make clear what the difference between this challenge and the previous one is, not for us to justify why. I added some suggestions for a goal section that I think addresses this.
